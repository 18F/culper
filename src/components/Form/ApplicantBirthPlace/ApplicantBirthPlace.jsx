import React from 'react'
import ValidationElement from '../validationElement'
import City from '../City'
import MilitaryState from '../MilitaryState'
import County from '../County'
import Country from '../Country'
import { api } from '../../../services/api'

export default class ApplicantBirthPlace extends ValidationElement {
  constructor (props) {
    super(props)

    this.state = {
      name: props.name,
      label: props.label,
      city: props.city,
      state: props.state,
      county: props.county,
      country: props.country
    }
  }

  /**
   * Handle the change event.
   */
  handleChange (event) {
    let part = this.extractPart(event.target.id)
    let value = event.target.value
    let updated = null

    switch (part) {
      case 'city':
        updated = { first: value }
        break

      case 'state':
        updated = { last: value }
        break

      case 'county':
        updated = { middle: value }
        break

      case 'country':
        updated = { suffix: value }
        break
    }

    if (updated !== null) {
      this.setState(updated, () => {
        super.handleChange(event)
      })
    } else {
      super.handleChange(event)
    }
  }

  /**
   * Handle the validation event.
   */
  handleValidation (event, status) {
    this.setState({error: status === false, valid: status === true}, () => {
      if (this.state.error === false || this.state.valid === true) {
        super.handleValidation(event, status)
        return
      }

      api
        .validateApplicantBirthplace({
          City: this.state.city,
          State: this.state.state,
          County: this.state.county,
          Country: this.state.country
        })
        .then((response) => {
          // TODO: Display and assign the errors as necessary
          if (response.Errors) {
          }

          if (response.Suggestions) {
          }
        })
        .then(() => {
          super.handleValidation(event, status)
        })
    })
  }

  /**
   * Generated name for the part of the address elements.
   */
  partName (part) {
    return '' + this.state.name + '-' + part
  }

  /**
   * Returns the part name from the pull generated identifier.
   */
  extractPart (id) {
    return id.split('-').pop()
  }

  render () {
    return (
      <div>
        <City name={this.partName('city')}
              label="City"
              value={this.state.city}
              onChange={this.handleChange}
              onValidate={this.handleValidation}
              onFocus={this.props.onFocus}
              onBlur={this.props.onBlur}
              />
        <MilitaryState name={this.partName('state')}
                       label="State"
                       value={this.state.state}
                       includeStates="true"
                       onChange={this.handleChange}
                       onValidate={this.handleValidation}
                       onFocus={this.props.onFocus}
                       onBlur={this.props.onBlur}
                       />
        <County name={this.partName('county')}
                label="County"
                value={this.state.county}
                onChange={this.handleChange}
                onValidate={this.handleValidation}
                onFocus={this.props.onFocus}
                onBlur={this.props.onBlur}
                />
        <Country name={this.partName('country')}
                 label="Country"
                 value={this.state.country}
                 onChange={this.handleChange}
                 onValidate={this.handleValidation}
                 onFocus={this.props.onFocus}
                 onBlur={this.props.onBlur}
                 />
      </div>
    )
  }
}
