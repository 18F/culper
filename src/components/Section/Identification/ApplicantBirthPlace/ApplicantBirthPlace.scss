.birthplace {
  .state,
  .city,
  .county,
  .country {
    display: inline-block;
    width: 90%;
    max-width: 64rem !important;

    input {
      width: 100% !important;
      max-width: 64rem !important;
    }
    padding-bottom: 2.2rem;
  }

  .branch {
    padding-top: 2rem;
    .help {
      padding-top: 0;

      label {
        position: relative;
      }
    }
  }
}
