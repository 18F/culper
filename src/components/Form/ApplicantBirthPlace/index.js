import ApplicantBirthPlace from './ApplicantBirthPlace'
export default ApplicantBirthPlace
