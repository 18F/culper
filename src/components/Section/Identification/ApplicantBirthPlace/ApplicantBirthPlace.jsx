import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import validate from '../../../../validators'
import SubsectionElement from '../../SubsectionElement'
import { Location, Field } from '../../../Form'

export default class ApplicantBirthPlace extends SubsectionElement {
  constructor (props) {
    super(props)
    this.update = this.update.bind(this)
    this.updateLocation = this.updateLocation.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      Location: this.props.Location,
      ...queue
    })
  }

  updateLocation (values) {
    this.update({
      Location: values
    })
  }

  render () {
    const klass = `applicant-birthplace ${this.props.className || ''}`.trim()

    return (
      <div className={klass}>
        <Field title={i18n.t('identification.birthplace.title')}
               titleSize="h2"
               scrollIntoView={this.props.scrollIntoView}>
          <Location name="birthplace"
                    {...this.props.Location}
                    layout={Location.BIRTHPLACE}
                    label={i18n.t('identification.birthplace.label.location')}
                    stateLabel={i18n.t('identification.birthplace.label.state')}
                    statePlaceholder={i18n.t('identification.birthplace.placeholder.state')}
                    cityLabel={i18n.t('identification.birthplace.label.city')}
                    cityPlaceholder={i18n.t('identification.birthplace.placeholder.city')}
                    countyLabel={i18n.t('identification.birthplace.label.county')}
                    countyPlaceholder={i18n.t('identification.birthplace.placeholder.county')}
                    countryLabel={i18n.t('identification.birthplace.label.country')}
                    countryPlaceholder={i18n.t('identification.birthplace.placeholder.country')}
                    onUpdate={this.updateLocation}
                    onError={this.handleError}
                    required={this.props.required}
                    />
        </Field>
      </div>
    )
  }
}

ApplicantBirthPlace.defaultProps = {
  location: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'birthplace',
  dispatch: () => {},
  validator: (data) => {
    return validate(schema('identification.birthplace', data))
  }
}

ApplicantBirthPlace.errors = []
