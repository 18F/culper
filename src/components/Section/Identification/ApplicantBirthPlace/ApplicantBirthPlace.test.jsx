import React from 'react'
import { shallow, mount } from 'enzyme'
import { ApplicantBirthPlace } from './ApplicantBirthPlace'

describe('The ApplicantBirthPlace component', () => {
  it('renders without errors', () => {
    const component = shallow(<ApplicantBirthPlace />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it('displays no error by default', () => {
    const component = mount(<ApplicantBirthPlace />)

    component.find('.blocks .yes input').simulate('change')
    expect(component.find('.usa-input-error-label').length).toEqual(0)
  })

  it('implements an onUpdate handler', () => {
    let updates = 0
    const testProps = {
      name: 'input-focus',
      className: 'input-focus',
      label: 'Text input focused',
      value: '',
      onUpdate: () => {
        updates += 1
      },
    }
    const component = mount(<ApplicantBirthPlace {...testProps} />)
    component.find('.blocks .yes input').simulate('change')
    expect(updates).toBe(1)
  })
})
