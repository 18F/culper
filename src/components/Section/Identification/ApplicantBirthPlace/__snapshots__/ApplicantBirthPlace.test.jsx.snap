// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ApplicantBirthPlace component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content applicant-birthplace"
    data-section="identification"
    data-subsection="birthplace"
  >
    <h1
      className="section-header"
    >
      Place of birth
    </h1>
    <Field
      adjustFor=""
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help=""
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Provide your place of birth"
      titleSize="h4"
      validate={true}
    >
      <Location
        addressBook=""
        addressBooks={Object {}}
        cityLabel="City"
        countryLabel="Country"
        countyLabel="County or province"
        dispatch={[Function]}
        geocode={false}
        geocodeResult={Object {}}
        label="Were you born in the United States?"
        layout="Birthplace"
        name="birthplace"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        spinner={false}
        stateLabel="State or territory"
        street2Label="Apt, suite, building, floor, etc."
        streetLabel="Street address"
        suggestions={false}
        validated={false}
        zipcodeLabel="ZIP Code"
      />
    </Field>
  </div>,
  "nodes": Array [
    <div
      className="section-content applicant-birthplace"
      data-section="identification"
      data-subsection="birthplace"
    >
      <h1
        className="section-header"
      >
        Place of birth
      </h1>
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help=""
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Provide your place of birth"
        titleSize="h4"
        validate={true}
      >
        <Location
          addressBook=""
          addressBooks={Object {}}
          cityLabel="City"
          countryLabel="Country"
          countyLabel="County or province"
          dispatch={[Function]}
          geocode={false}
          geocodeResult={Object {}}
          label="Were you born in the United States?"
          layout="Birthplace"
          name="birthplace"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          spinner={false}
          stateLabel="State or territory"
          street2Label="Apt, suite, building, floor, etc."
          streetLabel="Street address"
          suggestions={false}
          validated={false}
          zipcodeLabel="ZIP Code"
        />
      </Field>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ApplicantBirthPlace
        Location={Object {}}
        dispatch={[Function]}
        onError={[Function]}
        onUpdate={[Function]}
        validator={[Function]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ApplicantBirthPlace {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "Location": Object {},
          "dispatch": [Function],
          "onError": [Function],
          "onUpdate": [Function],
          "validator": [Function],
        },
        "refs": Object {},
        "section": "identification",
        "state": null,
        "store": "Identification",
        "storeKey": "ApplicantBirthPlace",
        "subsection": "birthplace",
        "update": [Function],
        "updateLocation": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content applicant-birthplace"
          data-section="identification"
          data-subsection="birthplace"
        >
          <h1
            className="section-header"
          >
            Place of birth
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your place of birth"
            titleSize="h4"
            validate={true}
          >
            <Location
              addressBook=""
              addressBooks={Object {}}
              cityLabel="City"
              countryLabel="Country"
              countyLabel="County or province"
              dispatch={[Function]}
              geocode={false}
              geocodeResult={Object {}}
              label="Were you born in the United States?"
              layout="Birthplace"
              name="birthplace"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              spinner={false}
              stateLabel="State or territory"
              street2Label="Apt, suite, building, floor, etc."
              streetLabel="Street address"
              suggestions={false}
              validated={false}
              zipcodeLabel="ZIP Code"
            />
          </Field>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="section-content applicant-birthplace"
          data-section="identification"
          data-subsection="birthplace"
        >
          <h1
            className="section-header"
          >
            Place of birth
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your place of birth"
            titleSize="h4"
            validate={true}
          >
            <Location
              addressBook=""
              addressBooks={Object {}}
              cityLabel="City"
              countryLabel="Country"
              countyLabel="County or province"
              dispatch={[Function]}
              geocode={false}
              geocodeResult={Object {}}
              label="Were you born in the United States?"
              layout="Birthplace"
              name="birthplace"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              spinner={false}
              stateLabel="State or territory"
              street2Label="Apt, suite, building, floor, etc."
              streetLabel="Street address"
              suggestions={false}
              validated={false}
              zipcodeLabel="ZIP Code"
            />
          </Field>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ApplicantBirthPlace
    Location={Object {}}
    dispatch={[Function]}
    onError={[Function]}
    onUpdate={[Function]}
    validator={[Function]}
  />,
}
`;
